Tests for torch device and dtype utils

DaniAffCH opened the pull request:
### Changes

This PR addresses #2579.
- Add tests for torch device utils. The tests consider the case in which the model has no parameters, has all parameters on CPU, has all parameters on CUDA, and has parameters placed on different devices. In the latter, the parameters are moved on different devices randomly. 
- Add tests for torch `dtype` utils. The case in which the model has no parameters is also considered.
- Created a test torch helper class `EmptyModel` to take into account the case in which the model has no parameters at all.
- Add docstrings in utils.

### Tests

I compared all the results manually checking for their correctness. The code is also compliant with the coding style having been verified with `pre-commit run`

daniil-lyakhov commented on tests/torch/test_utils.py:
```diff
@@ -83,3 +116,52 @@ def check_were_only_bn_training_state_changed(model: nn.Module, saved_state: _Mo
+        EmptyModel(),
+    ],
+)
+def test_model_device(model: nn.Module):
+    assert not is_multidevice(model)
+    assert get_model_device(model).type == "cpu"
+
+    multidevice = randomly_change_model_device(model)
+    assert is_multidevice(model) == multidevice
+
+    is_empty = is_model_empty(model)
+
+    if torch.cuda.is_available() and not is_empty:
+        cuda = torch.device("cuda")
+        model = model.to(cuda)
+        assert not is_multidevice(model)
+        assert get_model_device(model).type == "cuda"
```
Please simplify the test:
1) Do not use `randomly_change_model_device`, test should be 100% deterministic
2) Please test only two models: empty model and Two conv test model as they cover 100% of cases
3) Please split this test on two tests: `test_empty_model_device` and `test_model_device` and remove this if section.

I expect following tests here:
`test_empty_model_device`: test creates empty model and check device
`test_model_device` skipped if cuda is not available, creates model on cpu, checks device and multidevice, move one param to cude, check devce and multidevce, move all params to cude, check device and multidevice

daniil-lyakhov commented on tests/torch/test_utils.py:
```diff
@@ -83,3 +116,52 @@ def check_were_only_bn_training_state_changed(model: nn.Module, saved_state: _Mo
+    ],
+)
+def test_model_dtype(model: nn.Module):
+    is_empty = is_model_empty(model)
+
+    model.to(torch.float16)
+    assert is_empty or get_model_dtype(model) == torch.float16
+    model.to(torch.float32)
+    assert get_model_dtype(model) == torch.float32
+    model.to(torch.float64)
+    assert is_empty or get_model_dtype(model) == torch.float64
```
Please split this test on two:
`test_empty_model_dtype` and `test_model_dtype`

daniil-lyakhov reviewed (changes requested):
@DaniAffCH, thank you for the contribution! Please address comments below

DaniAffCH commented:
Hi, @daniil-lyakhov thank you for the review, I just addressed your comments.

While refactoring I noticed that there are some tests that are not deterministic, previously developed. 
For instance:  https://github.com/openvinotoolkit/nncf/blob/e76bba2f8039266a1af42a01b1bbbcdb79ebc1d5/tests/torch/test_utils.py#L40

Should I modify them to ensure a deterministic outcome? To not modify it too much I could set a fixed random seed

daniil-lyakhov commented on tests/torch/test_utils.py:
```diff
@@ -83,3 +87,46 @@ def check_were_only_bn_training_state_changed(model: nn.Module, saved_state: _Mo
     compare_saved_model_state_and_current_model_state(model, saved_state)
+
+
+def test_model_device():
+    if not torch.cuda.is_available():
+        return
+
+    model = TwoConvTestModel()
+    cuda = torch.device("cuda")
```
```suggestion
@pytest.mark.skipif(not torch.cuda.is_available(), reason="Cuda is not available in current environment")
def test_model_device():
    model = TwoConvTestModel()
    cuda = torch.device("cuda")
```

daniil-lyakhov reviewed (commented):
Minor

daniil-lyakhov commented:
> some other tests that

Nice catch! looks like we missed this part during code review previously. This test could cause problems in future because of randomness, it would be great if you remove randomness from them

DaniAffCH commented:
> > some other tests that
> 
> Nice catch! looks like we missed this part during code review previously. This test could cause problems in future because of randomness, it would be great if you remove randomness from them

Perfect, done! :)

DaniAffCH commented on tests/torch/test_utils.py:
```diff
@@ -76,10 +72,51 @@ def check_were_only_bn_training_state_changed(model: nn.Module, saved_state: _Mo
     runner = DataLoaderBNAdaptationRunner(model, "cuda")
 
-    randomly_change_model_state(model)
+    change_model_state(model)
```
Do you mean just inside the test `test_bn_training_state_switcher` or also in `test_training_mode_switcher`? In case both has to be modified I can simply change the `change_model_state` function

daniil-lyakhov replied:
Please do it only in the `test_bn_training_state_switcher`, `test_training_mode_switcher` is ok

daniil-lyakhov commented on tests/torch/test_utils.py:
```diff
@@ -76,10 +72,51 @@ def check_were_only_bn_training_state_changed(model: nn.Module, saved_state: _Mo
     runner = DataLoaderBNAdaptationRunner(model, "cuda")
 
-    randomly_change_model_state(model)
+    change_model_state(model, trainable_paramaters=False)
```
```suggestion
    for p in module.parameters():
        p.requires_grad = False
```

daniil-lyakhov commented on tests/torch/test_utils.py:
```diff
@@ -33,29 +36,22 @@ def compare_saved_model_state_and_current_model_state(model: nn.Module, model_st
 
-def randomly_change_model_state(module: nn.Module, compression_params_only: bool = False):
-    import random
+def change_model_state(module: nn.Module, trainable_paramaters: bool = True):
```
Please remove `trainable_paramaters` param
```suggestion
def change_model_state(module: nn.Module):
```

daniil-lyakhov reviewed (approved):
LGTM 👍

daniil-lyakhov commented:
@AlexanderDokuchaev, please review and merge if you don't have comments
